File: Backend/password_policy.py
"""Password strength validation for registration and password changes."""
import re

MIN_PASSWORD_LENGTH = 10
MAX_PASSWORD_LENGTH = 128

_PASSWORD_RULES = [
    (r"[a-z]", "one lowercase letter"),
    (r"[A-Z]", "one uppercase letter"),
    (r"\d", "one number"),
    (r"[^\w\s]", "one special character"),
]


def validate_password_strength(password: str) -> None:
    if not password or not isinstance(password, str):
        raise ValueError("Password is required.")

    if len(password) < MIN_PASSWORD_LENGTH:
        raise ValueError(f"Password must be at least {MIN_PASSWORD_LENGTH} characters.")

    if len(password) > MAX_PASSWORD_LENGTH:
        raise ValueError(f"Password must be at most {MAX_PASSWORD_LENGTH} characters.")

    missing = [label for pattern, label in _PASSWORD_RULES if not re.search(pattern, password)]
    if missing:
        raise ValueError(
            "Password must include " + ", ".join(missing) + "."
        )

    if password.strip() != password:
        raise ValueError("Password cannot start or end with spaces.")

File: Frontend/src/utils/passwordPolicy.js
export const PASSWORD_MIN_LENGTH = 10;

export function passwordRequirements(password = "") {
    return [
        { key: "length", label: `At least ${PASSWORD_MIN_LENGTH} characters`, met: password.length >= PASSWORD_MIN_LENGTH },
        { key: "lower", label: "One lowercase letter", met: /[a-z]/.test(password) },
        { key: "upper", label: "One uppercase letter", met: /[A-Z]/.test(password) },
        { key: "digit", label: "One number", met: /\d/.test(password) },
        { key: "special", label: "One special character", met: /[^\w\s]/.test(password) },
    ];
}

export function isPasswordStrong(password = "") {
    return passwordRequirements(password).every((r) => r.met) && password.trim() === password;
}

export function passwordValidationMessage(password = "") {
    if (!password) return "Password is required.";
    if (password.trim() !== password) return "Password cannot start or end with spaces.";
    const unmet = passwordRequirements(password).filter((r) => !r.met);
    if (!unmet.length) return null;
    return `Password must include ${unmet.map((r) => r.label.toLowerCase()).join(", ")}.`;
}
